p2p/src/test/kotlin/maru/p2p/P2PTest.kt

```diff
@@ -0,0 +1,387 @@
+    val p2pManager2 = P2PManager(key2, P2P(IPV4, PORT2, listOf(PEER_ADDRESS_NODE_1)))
+    try {
+      p2pManager1.start()
+      val p2pNetwork1 = p2pManager1.p2pNetwork
+      p2pNetwork1.subscribe("topic", TestTopicHandler())
+
+      p2pManager2.start()
+      val p2pNetwork2 = p2pManager2.p2pNetwork
+      val testTopicHandler2 = TestTopicHandler()
+      p2pNetwork2.subscribe("topic", testTopicHandler2)
+
+      Awaitility
+        .await()
+        .timeout(
+          6000,
+          TimeUnit.MILLISECONDS,
+        ).untilAsserted { assertThat(p2pNetwork1.peerCount).isEqualTo(1) }
+      Awaitility
+        .await()
+        .timeout(
+          6000,
+          TimeUnit.MILLISECONDS,
+        ).untilAsserted { assertThat(p2pNetwork2.peerCount).isEqualTo(1) }
+
+      p2pNetwork1.gossip("topic", Bytes.fromHexString(GOSSIP_MESSAGE))
+
+      assertThat(
+        testTopicHandler2.dataFuture.get(10, TimeUnit.MILLISECONDS),
+      ).isEqualTo(Bytes.fromHexString(GOSSIP_MESSAGE))
+    } finally {
+      p2pManager1.stop()
+      p2pManager2.stop()
+    }
+  }
+
+  @Test
+  fun `peer receiving gossip passes message on`() {
+    val p2pManager1 = P2PManager(key1, P2P(IPV4, PORT1, emptyList()))
+    val p2pManager2 =
+      P2PManager(key2, P2P(IPV4, PORT2, listOf(PEER_ADDRESS_NODE_1, PEER_ADDRESS_NODE_3)))
+    val p2pManager3 = P2PManager(key3, P2P(IPV4, PORT3, emptyList()))
+    try {
+      p2pManager1.start()
+      val p2pNetwork1 = p2pManager1.p2pNetwork
+      p2pNetwork1.subscribe("topic", TestTopicHandler())
+
+      p2pManager2.start()
+      val p2pNetwork2 = p2pManager2.p2pNetwork
+      val testTopicHandler2 = TestTopicHandler()
+      p2pNetwork2.subscribe("topic", testTopicHandler2)
+
+      p2pManager3.start()
+      val p2pNetwork3 = p2pManager3.p2pNetwork
+      val testTopicHandler3 = TestTopicHandler()
+      p2pNetwork3.subscribe("topic", testTopicHandler3)
+
+      Awaitility
+        .await()
+        .timeout(
+          6000,
+          TimeUnit.MILLISECONDS,
+        ).untilAsserted {
+          assertThat(
+            p2pNetwork1.isConnected(
+              PeerAddress(
+                LibP2PNodeId(
+                  PeerId.fromBase58(
+                    PEER_ID_NODE_2,
+                  ),
+                ),
+              ),
+            ),
+          ).isTrue()
+        }
+      Awaitility
+        .await()
+        .timeout(
+          6000,
+          TimeUnit.MILLISECONDS,
+        ).untilAsserted {
+          assertThat(
+            p2pNetwork3.isConnected(
+              PeerAddress(
+                LibP2PNodeId(
+                  PeerId.fromBase58(
+                    PEER_ID_NODE_2,
+                  ),
+                ),
+              ),
+            ),
+          ).isTrue()
+        }
+      assertThat(p2pNetwork1.peerCount).isEqualTo(1)
+      assertThat(p2pNetwork2.peerCount).isEqualTo(2)
+      assertThat(p2pNetwork3.peerCount).isEqualTo(1)
```

Comment:
From the peer counts, it follows that p2pNetwork1 and 3 are probably not connected directly, but we could use a direct check

Reply:
The two awaitability calls on top of that are making sure that 1 is connected to 2 and 3 is connected to 2, these 3 lines make sure that these connections are all we have.